auth/auth.go

```diff
@@ -0,0 +1,281 @@
+				Cause:      errors.Wrap(err, "set state"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		// NOTE: Using same value for both state and nonce is fine because "state" is
+		// for service self-validation (no CSRF) and "nonce" is for IdP validation (no
+		// repeated attempts).
+		params := url.Values{}
+		params.Add("client_id", h.config.ClientID)
+		params.Add("redirect_uri", h.config.RedirectURI)
+		params.Add("state", nonce)
+		params.Add("nonce", nonce)
+		params.Add("response_type", "code")
+		params.Add("scope", strings.Join(scopes.ToStrings(h.config.RequestScopes), " "))
+
+		// Passthrough the IdP-aware query parameters.
+		params.Add("prompt", r.URL.Query().Get("prompt"))
+		params.Add("prompt_auth", r.URL.Query().Get("prompt_auth"))
+
+		http.Redirect(w, r, p.Endpoint().AuthURL+"?"+params.Encode(), http.StatusFound)
+	})
+}
+
+// CallbackHandler returns an HTTP handler that handles the SAMS callback and
+// calls the success handler upon successful authentication. Use
+// UserInfoFromContext to extract the user information.
+func (h *Handler) CallbackHandler(success http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		nonce, err := h.config.StateGetter(r)
+		h.config.StateDeleter(w, r) // Delete the state after getting it to make sure it's one-time use.
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      errors.Wrap(err, "get state"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+		if got := r.URL.Query().Get("state"); nonce != got {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusBadRequest,
+				Cause:      errors.Errorf("mismatched state, want %q but got %q", nonce, got),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		userInfo, err := h.getUserInfo(r)
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      err,
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		ctx = WithUserInfo(ctx, userInfo)
+		success.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// DefaultFailureHandler responds with the status code and message based on the
+// error extracted from the context.
+var DefaultFailureHandler = http.HandlerFunc(failureHandler)
+
+func failureHandler(w http.ResponseWriter, r *http.Request) {
+	err := ErrorFromContext(r.Context())
+	http.Error(w, err.Cause.Error(), err.StatusCode)
+}
+
+// randomString returns a base64-encoded random 32 byte string.
+func randomString() (string, error) {
+	b := make([]byte, 32)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", err
+	}
+	return base64.RawURLEncoding.EncodeToString(b), nil
+}
+
+// UserInfo contains the information about the authenticated user.
+type UserInfo struct {
```

Comment:
Where does this come from? Is it from the OIDC or OAuth spec? Or is it specific to the SAMS system? It would be nice to call that out, since if it were the latter we'd need to generally keep it in-sync with any other fields we add to the SAMS gRPC API, right?

Reply:
It is mapping to whatever the SAMS discovery endpoint (1) said the "userinfo" endpoint's (2) response body is.

![CleanShot 2024-04-01 at 16 41 48@2x](https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/assets/2946214/8655d2d8-35b1-49d0-bd60-fd392360083d)

All of fields included in this PR are from OIDC spec, the server-side counterpart https://github.com/sourcegraph/sourcegraph-accounts/blob/c40d8f0d8374bb2222124f87f4f3626c51c4bd2e/backend/internal/api/model.go#L11-L17